go/fixchain/fix.go

```diff
@@ -0,0 +1,86 @@
+package fixchain
+
+import (
+	"encoding/pem"
+	"log"
+
+	"github.com/google/certificate-transparency/go/x509"
+)
+
+type toFix struct {
+	cert  *x509.Certificate
+	chain *DedupedChain
+	opts  *x509.VerifyOptions
+	fixer *Fixer
+}
+
+func (fix *toFix) handleChain() ([][]*x509.Certificate, *FixError) {
+	chains, ferr := fix.constructChain()
+	if ferr != nil {
+		fix.fixer.errors <- ferr
+		chains, ferr = fix.fixChain()
+	}
+	return chains, ferr
+}
+
+func (fix *toFix) constructChain() ([][]*x509.Certificate, *FixError) {
+	chains, err := fix.cert.Verify(*fix.opts)
+	if err != nil {
+		fix.fixer.notReconstructed++
+		return chains, &FixError{Type: VerifyFailed, Cert: fix.cert,
+			Chain: fix.chain.certs, Error: err}
+	}
+	fix.fixer.reconstructed++
+	return chains, nil
+}
+
+func (fix *toFix) fixChain() ([][]*x509.Certificate, *FixError) {
+	d2 := *fix.chain
+	d2.AddCert(fix.cert)
+	for _, c := range d2.certs {
+		urls := c.IssuingCertificateURL
+		for _, url := range urls {
+			fix.augmentIntermediates(url)
+			chains, err := fix.cert.Verify(*fix.opts)
+			if err == nil {
+				fix.fixer.fixed++
+				return chains, nil
+			}
+		}
+	}
+	fix.fixer.notFixed++
+	return nil, &FixError{Type: FixFailed, Cert: fix.cert,
+		Chain: fix.chain.certs}
+}
+
+func (fix *toFix) augmentIntermediates(url string) {
+	// PKCS#7 additions
```

Comment:
Perhaps note that these exist because I couldn't find a go PKCS#7 library I wanted to use.